sdk/spring-sdk/src/main/scala/kalix/springsdk/impl/ProtoDescriptorGenerator.scala

```diff
@@ -0,0 +1,196 @@
+      }
+
+    private def buildMessageType(
+        protobufSchema: ProtobufSchema,
+        messageType: Class[_]): DescriptorProtos.DescriptorProto = {
+
+      // FIXME: still need to find out how is this structured.
+      // I guess we will need to traverse the type looking for nested types (non scalar types)
+      val fields = protobufSchema.getRootType.fields().asScala.map { field =>
+
+        val builder =
+          DescriptorProtos.FieldDescriptorProto.newBuilder
+            .setName(field.name)
+            .setNumber(field.id)
+            .setType(mapScalarType(field.`type`))
+
+        builder.build()
+      }
+
+      DescriptorProtos.DescriptorProto.newBuilder
+        .setName(protobufSchema.getRootType.getName)
+        .addAllField(fields.asJava)
+        .build()
+    }
+
+    def genFileDescriptor(
+        name: String,
+        packageName: String,
+        handlers: Seq[Method],
+        typesSelector: Method => TypesSelector): Descriptors.FileDescriptor = {
+
+      val protoMapper = new ProtobufMapper
+
+      val protoBuilder = DescriptorProtos.FileDescriptorProto.newBuilder
+      protoBuilder
+        .setName(name + ".proto") // FIXME: snake_case this ?!
+        .setSyntax("proto3")
+        .setPackage(packageName)
+        .setOptions(DescriptorProtos.FileOptions.newBuilder.setJavaMultipleFiles(true).build)
+
+      // build messages types
+      handlers
+        .flatMap(method => typesSelector(method).allMessageTypes)
+        // types can be used many times, we need unique descriptors
+        .toSet
+        .foreach { messageType: Class[_] =>
+          val schema = protoMapper.generateSchemaFor(messageType)
+          protoBuilder.addMessageType(buildMessageType(schema, messageType))
+        }
+
+      // build gRPC service
+      val serviceBuilder = DescriptorProtos.ServiceDescriptorProto.newBuilder
+      serviceBuilder.setName(name)
+
+      handlers.foreach { method =>
+        val methodBuilder = DescriptorProtos.MethodDescriptorProto.newBuilder
+
+        val selector = typesSelector(method)
+        val input = selector.inputType.getName
+        val output = selector.outputType.getName
+
+        methodBuilder
+          .setName(method.getName.capitalize)
+          .setInputType(input)
+          .setOutputType(output)
+
+        serviceBuilder.addMethod(methodBuilder.build())
+      }
+
+      protoBuilder.addService(serviceBuilder.build())
+
+      // finally build all final descriptor
+      Descriptors.FileDescriptor.buildFrom(protoBuilder.build, new Array[Descriptors.FileDescriptor](0))
+    }
+  }
+
+  def generateFileDescriptorAction(component: Class[_ <: Action]): Descriptors.FileDescriptor = {
+    val handler =
+      component.getDeclaredMethods
+        .filter(_.getReturnType == classOf[Action.Effect[_]])
+        // actions have only one input param, always
+        .filter(_.getParameters.length == 1)
```

Comment:
Should perhaps also filter non public methods, or will this filtering be replaced by looking for some resty method annotation to decide if something should be in the descriptor?